frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx

```diff
@@ -461,6 +505,67 @@ class SegmentsView extends React.Component<Props, State> {
     }
   }
 
+  onExpandTree = (expandedKeys: Key[]) => {
+    this.expandGroups(expandedKeys);
+  };
+
+  getSubGroupsAsTreeNodes = (groupId: number) => {
+    if (groupId !== MISSING_GROUP_ID) {
+      return getGroupByIdWithSubgroups(this.props.segmentGroups, groupId)
+        .filter((group) => group !== groupId)
+        .map((group) => getKeyForGroupId(group));
+    }
+    const allSegmentGroups = this.props.segmentGroups.flatMap((group) =>
+      getGroupByIdWithSubgroups(this.props.segmentGroups, group.groupId),
+    );
+    return allSegmentGroups.map((group) => getKeyForGroupId(group));
+  };
+
+  expandGroups = (newExpandedGroups: Key[]) => {
+    if (this.props.visibleSegmentationLayer == null) return;
+    const expandedKeySet = new Set(newExpandedGroups);
+    if (this.state.isRootGroupExpanded && !expandedKeySet.has(getKeyForGroupId(MISSING_GROUP_ID))) {
+      this.setState({ isRootGroupExpanded: false });
+    } else if (
+      !this.state.isRootGroupExpanded &&
+      expandedKeySet.has(getKeyForGroupId(MISSING_GROUP_ID))
+    ) {
+      this.setState({ isRootGroupExpanded: true });
+    }
+    const newGroups = mapGroups(this.props.segmentGroups, (group) => {
+      if (expandedKeySet.has(getKeyForGroupId(group.groupId))) {
+        return {
+          ...group,
+          isExpanded: true,
+        };
+      } else {
+        return {
+          ...group,
+          isExpanded: false,
+          // Close all groups that are not in the expanded list so this method
+          // can be called for every update, e.g. when a group is collapsed.
+        };
+      }
+    });
+    Store.dispatch(setSegmentGroupsAction(newGroups, this.props.visibleSegmentationLayer?.name));
+  };
+
+  collapseGroups = (groupsToCollapse: Key[]) => {
```

Comment:
why doesn't this method handle the root group like the `expandGroups` method does? is it not necessary in this case?

Reply:
also: the two methods differ in their semantics of the parameters. for `collapseGroups` I understand it so that the referenced groups will be collapsed (and other collapsed groups will stay as before). for `expandGroups`, it seems like only the referenced groups should be expanded (and all the others should not be expanded). can this be unified?

Reply:
you're right, the methods differ. this is because `expandGroups` is called by the Tree component _and_ the `expandSubgroups` menu entry. The tree component also calls this if groups are collapsed, by removing the collapsed groups trees from the list. Thus `isExpanded` has to be set to false for these groups. Other than that, the root group is handled differently because on `expand all subgroups` the parent group is expanded if it was collapsed before. This is different for expanded groups, their state doesn't change if their children are collapsed.
I could use three methods: one for the Tree component and two for "collapse/expand subgroups". And then comment what I said above. Do you think that this might be a good solution? @philippotto

Reply:
ah, I understand. two methods are fine then. I'd the `expandGroups` method to `setExpandedGroups`. That way, it should be clearer that the semantics is different to `collapseGroups`.

> Other than that, the root group is handled differently because on expand all subgroups the parent group is expanded if it was collapsed before. This is different for expanded groups, their state doesn't change if their children are collapsed.

A comment for that would be good, yes :+1: